proof/refine/ARM/Ipc_R.thy

```diff
@@ -5199,12 +5265,180 @@ lemma cancelIPC_notin_epQueue:
   unfolding cancelIPC_def
   by (wpsimp wp: blockedCancelIPC_notin_epQueue hoare_drop_imps threadSet_valid_objs')
 
+crunches rescheduleRequired
+  for scs_tcbs_of[wp]: "\<lambda>s. P (tcb_scs_of' s) (scTCBs_of s)"
+  (wp: crunch_wps threadSet_tcb_scs_of'_inv ignore: threadSet)
+
+lemma maybeReturnSc_tcbs_scs_sym_refs[wp]:
+  "\<lbrace>tcbs_scs_sym_refs and valid_objs'\<rbrace>
+   maybeReturnSc y t
+   \<lbrace>\<lambda>_. tcbs_scs_sym_refs\<rbrace>"
+  unfolding maybeReturnSc_def
+  apply (simp add: liftM_def)
+  apply (rule hoare_seq_ext[OF _ stateAssert_sp])
+  apply (rule hoare_seq_ext[OF _ get_ntfn_sp'])
+  apply wpsimp
+      apply (wps, wpsimp wp: setSchedContext_scTCBs_of)
+     apply wpsimp+
+    apply (wps, wpsimp wp: threadSet_tcb_scs_of')
+   apply (wpsimp wp: threadGet_wp)
+  apply (clarsimp simp: tcb_at'_ex_eq_all)
+  apply (drule sym, simp)
+  apply (subgoal_tac "pred_map_eq (the (tcbSchedContext tcb)) (tcb_scs_of' s) t")
+   apply (clarsimp simp: sqheap_refs_inv_defs pred_map_eq_upd)
+   apply (auto)[1]
```

Comment:
Does something else work here? If not, maybe change to `subgoal by auto`? (I think it'd be good if we could decide on what to do with these `auto`s)